src/Device/AndroidSensors.cpp

```diff
@@ -292,18 +292,21 @@ DeviceDescriptor::OnEngineSensors(bool has_cht,
                                   Temperature cht,
                                   bool has_egt,
                                   Temperature egt,
-                                  bool has_ignitions_per_sec,
-                                  float ignitions_per_sec) noexcept
+                                  bool has_ignitions_per_second,
+                                  float ignitions_per_second) noexcept
 {
   const auto e = BeginEdit();
   NMEAInfo &basic = *e;
   basic.UpdateClock();
   basic.alive.Update(basic.clock);
-  if(has_ignitions_per_sec){
-    basic.engine_state.ignitions_per_sec = ignitions_per_sec;
-    basic.engine_state.ignitions_per_sec_available.Update(basic.clock);
+  if(has_ignitions_per_second){
+    /** Workaround, because we use index of DataFieldEnum in DeviceEditWidget for scaling,
+     *  which is int, but we need scaling of 0.5, 1.0 and 2.0.
+    */
```

Comment:
The API documentation of scale_to_rps doesn't explain this, but it should.
My idea, however, was to have "symbolic" values, e.g.:
```
enum class EngineType {
  TWO_STROKE_1_IGN,
  TWO_STROKE_2_IGN,
  FOUR_STROKE_1_IGN,
};
```
... but that's truly ugly. Reading your latest code again, maybe we should have two fields:
- engine type (2 = 2-stroke, 4 = 4-stroke)
- ignitions per revolution (1 or 2)

Does this concept scale well with all kinds of engines that should ever be supported by XCSoar? (I have really no idea of engines. I just want the software representation to be as simple and scalable as possible. Using weird fixed-point scaling that is not properly documented scares me a bit.)

Reply:
Yea, I don't know either what types of engines will get supported, but it will get worse with multi cylinder. 2 Cylinder V configuration is different to inline configuration, for both 2 and 4 stroke.
I will properly document the fixed-point stuff, to me this seams the easiest approach. The engine related parts should also be documented in the manual, right? How to choose the right scaling for the used engine...

Reply:
> maybe we should have two fields

Seems to much to me, but at the end, you decide :)

Reply:
> it will get worse with multi cylinder

Then... let's make it as simple as possible for users, and go with the `enum class EngineType` with a set of hard-coded choices. The first enum choice (the one that will get assigned the `0` value by the compiler) should be the one that applies no scaling at all (factor = `1`).

This solution is ugly, but on the other hand is extensible easily (if new engine sensor products get added), is robust, and easy to use for users. Robustness & ease of use are really important, and I think that outweighs the ugliness in this particular case.

Your suggestion to use fixed-point integers can be fragile and not so easily extensible. The numeric representation has a meaning, unlike my `enum EngineType` suggestion, but that's not a real advantage here, it will never materialize.

Reply:
So I add a 

```
enum class EngineType : unsigned{
  TWO_STROKE_1_IGN = 10,
  TWO_STROKE_2_IGN = 5,
  FOUR_STROKE_1_IGN = 20,
};
```

use it like

```
static void
FillRPMScaling(DataFieldEnum &dfe) noexcept
{
  dfe.addEnumText(_T("2S1I"), TWO_STROKE_1_IGN);
  dfe.addEnumText(_T("2S2I"), TWO_STROKE_2_IGN);
  dfe.addEnumText(_T("4S1I"), FOUR_STROKE_1_IGN);
}
```

change `unsigned scale_to_rps` in `Config.hpp` to `EngineType engine_type` and finally have a method in `AndroidSensors.cpp` `float EngineTypeToScalingFactor(EngineType engine_type)` which does the scaling?

Reply:
Omit those integers (=10, =5, =20); don't pretend they have a meaning. They are just arbitrary identifiers.
If you use automatically ascending integers, then the scaling factors can be stored in a simple array (right below the `enum` definition?):
```
constexpr float ignitions_to_revs_factors[] = { 1, 0.5, 2 };
constexpr float
IgnitionsToRevs(float ignitions, EngineType type) noexcept
{
  return ignitions * ignitions_to_revs_factors[static_cast<std::size_t>(type)];
}
```

Reply:
`enum class EngineType : unsigned{`
Better: `: uint_least8_t` - in the foreseeable future, we won't have more than 256 engine types ;-)

Reply:
Okay sounds good. No `if` involved, I like :)

`std::map<EngineType, float>` would be to big of a gun for the problem?

Reply:
I thought it of mapping a enum to a float... but little did I know about the obvious bad implementation of map...

Reply:
`std::map` is an excellent container type with a good implementation, and XCSoar uses it a lot. But as with every data type, it is not universal; it has characteristics that are good for one use case, but bad for others. For your code, it would be an extremely bad choice.
"If all you have is a hammer, everything looks like a nail" - when you write high-level code, don't expect everything to be abstracted away from you. There is no universal data type for everything.

Reply:
Oh, I see. Just ignore type safety like `enum class PressureUse` does. 
I guess at some point you have to compromise...

Reply:
`DataFieldEnum` is not a template. It could be, but then every use would generate a new implementation. To avoid this, we `DataFieldEnum` does not actually use C++ enums, but plain integers. This needs to be casted, where you leave type-safe terrain. That's okay, and still better than being type-unsafe everywhere.